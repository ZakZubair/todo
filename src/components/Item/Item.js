import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import checkIn from '../../assets/img/check-in.svg';

class Item extends PureComponent {
  render() {
    const {
      id, name, description, date, checked, onToggle, onRemove, onEdit,
    } = this.props;

    return (
      <div className="item" onClick={() => onToggle(id)} onKeyPress={() => onToggle(id)} role="button" tabIndex={id}>
        <div className="item__check-mark">
          { checked && (
          <img src={checkIn} alt="checked" />
          ) }
        </div>
        <div className={`item__text ${checked ? 'item__checked' : ''}`}>
          <p className="item__text__name"><span>{name}</span></p>
          <p className="item__text__description">{description}</p>
          <p className="item__text__date">{`Created at ${date}`}</p>
        </div>
        <button
          className="item__edit"
          onClick={(e) => {
            e.stopPropagation();
            onEdit(id);
          }}
          type="button"
        >
          Edit
        </button>
        <button
          className="item__remove"
          onClick={(e) => {
            e.stopPropagation();
            onRemove(id);
          }}
          type="button"
        >
          &times;
        </button>
      </div>
    );
  }
}

Item.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  checked: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
};
export default Item;
