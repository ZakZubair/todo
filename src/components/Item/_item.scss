.item {
  padding: rem(16);
  display: flex;
  align-items: center;
  transition: all 0.15s;
  user-select: none;
  outline: 0;
  border-top: rem(2) solid $athens-gray;
  cursor: pointer;

  &:hover {
    background: $emerald-20;

    .item__edit, .item__remove {
      opacity: 1;
    }
  }

  + .item {
    border-top: rem(2) solid $athens-gray;
  }

  &__text {
    flex: 1;
    word-break: break-all;
    margin-left: rem(16);
    transition: all 0.15s;

    &__name {
      font-size: rem(18);
      font-weight: 700;
    }

    &__description {
      font-size: rem(16);
      padding: rem(8) 0 0;
    }

    &__date {
      font-size: rem(8);
      padding-top: rem(8);
    }

    &__name, &__description, &__date  {
      color: $bright-gray;
      margin: 0;
    }

    input {
      width: 100%;
      color: $bright-gray;
      outline: none;
      border: none;
      padding-bottom: rem(4);
      border-bottom: rem(1) solid $french-gray;
      background-color: transparent;
      &.name {
        font-size: rem(18);
        font-weight: 700;
      }
      &.description {
        font-size: rem(16);
        padding: rem(8) 0 rem(4);
      }
    }
  }

  @keyframes strike{
    0%   { width : 0; }
    100% { width: 100%; }
  }

  &__checked {
    .item__text__name {
      span {
        position: relative;
        &:after {
          content: '';
          position: absolute;
          top: 50%;
          left: 0;
          width: 100%;
          height: rem(2);
          background: $french-gray;
          animation-name: strike;
          animation-duration: 0.2s;
          animation-timing-function: linear;
          animation-iteration-count: 1;
          animation-fill-mode: forwards;
        }
      }
    }
    .item__text {
      &__name, &__description, &__date {
        color: $french-gray;
      }
    }
  }

  &__check-mark {
    font-family: $secondary-font;
    font-size: rem(24);
    width: rem(32);
    height: rem(32);
    background-color: $white;
    color: $turquoise;
    border: rem(2) solid $athens-gray;
    transition: all 0.15s;
    @include border-radius(100%);

    img {
      width: rem(20);
      height: rem(20);
      padding-left: rem(6);
    }
  }

  &__edit {
    color: $bright-gray;
    opacity: 0;
    background: transparent;
    border: none;
    padding: 0 rem(16);
    outline: 0;
    transition: all 0.15s;
    cursor: pointer;
    &:hover {
      text-decoration: underline;
    }
  }

  &__remove {
    font-size: rem(16);
    font-family: $secondary-font;
    color: $emerald;
    font-weight: 700;
    opacity: 0;
    width: rem(28);
    height: rem(28);
    border: rem(2) solid $athens-gray;
    background-color: $white;
    transition: all 0.15s;
    outline: 0;
    cursor: pointer;
    @include border-radius(100%);

    &:hover {
      background-color: $emerald;
      color: white;
    }
  }
}
