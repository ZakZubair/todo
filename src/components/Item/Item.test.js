import 'jsdom-global/register';
import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Item from './Item';

const mockCallBack = jest.fn();
const props = {
  id: 'abc123',
  name: 'Test name',
  description: 'Test description',
  date: 'Date time',
  checked: false,
  onToggle: mockCallBack,
  onRemove: mockCallBack,
  onEdit: mockCallBack,
};

describe('test Item component', () => {
  it('render correctly item component', () => {
    const ItemComponent = renderer.create(<Item {...props} />).toJSON();
    expect(ItemComponent).toMatchSnapshot();
  });

  it('render item correctly with values', () => {
    const ItemComponent = mount(<Item {...props} />);
    expect((ItemComponent).prop('name')).toEqual('Test name');
    expect((ItemComponent).prop('description')).toEqual('Test description');
    expect((ItemComponent).prop('date')).toEqual('Date time');
    expect((ItemComponent).prop('checked')).toEqual(false);
    expect(ItemComponent.state().isEditing).toBeFalsy();
  });

  it('check the item is not editing', () => {
    const ItemComponent = mount(<Item {...props} />);
    expect(ItemComponent.state().isEditing).toBeFalsy();
  });

  it('check the item is editing', () => {
    const ItemComponent = mount(<Item {...props} />);
    ItemComponent.setState({ isEditing: true });
    expect(ItemComponent.state().isEditing).toBeTruthy();
  });
});
