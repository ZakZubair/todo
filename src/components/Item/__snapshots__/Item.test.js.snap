// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Item component render correctly item component 1`] = `
<div
  className="item"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="abc123"
>
  <div
    className="item__check-mark"
  />
  <div
    className="item__text "
  >
    <p
      className="item__text__name"
    >
      <span>
        Test name
      </span>
    </p>
    <p
      className="item__text__description"
    >
      Test description
    </p>
    <p
      className="item__text__date"
    >
      Created at Date time
    </p>
  </div>
  <button
    className="button__edit"
    onClick={[Function]}
    type="button"
  >
    Edit
  </button>
  <button
    className="button__remove"
    onClick={[Function]}
    type="button"
  >
    Ã—
  </button>
</div>
`;
